lib/webpacker/commands.rb

```diff
@@ -37,14 +59,16 @@ def versions
       manifest_config = Dir.glob("#{config.public_manifest_path}*")
 
       packs = all_files - manifest_config - current_version
-      packs.group_by { |file| File.mtime(file).utc.to_i }.sort.reverse
+      packs.reject { |file| File.directory?(file) }.group_by { |file| File.mtime(file).utc.to_i }
```

Comment:
This still assumes that all the files for a given compilation run of webpack are going to be outputted within 1 second, and so can be grouped into distinct runs and counted that way. That's not a valid assumption. 

The max_age code helps, but this still could lead to confusing and incorrect behavior. If files are older than the `age` param, they could still get grouped incorrectly and deleted prematurely. 

Also, with default capistrano config or heroku buildpacks or docker deploys, users don't often have the option of (easily) changing the params given to `rake assets:clean`, or passing that param along to the `webpack:clean` task. So it's probably still important to make sure that the `count` method is reliable.

Reply:
@sbeam see my comment above. How about cleaning before compilation, rather than after with all this logic.

Reply:
@justin808 that would work for some situations (including mine), but as I understand it, the point of keeping the last 2-3 versions of each compiled asses is in cases of rolling deploys or clients that have partially cached some of them, and there are still references to them out in the wild. So suddenly 404ing all old assets on each deploy isn't good. This is potentially an issue if you have e.g. capistrano deploys to multiple servers behind a naive load balancer that complete at different times.

That's why sprockets has [similar code](https://github.com/rails/sprockets/blob/master/lib/sprockets/manifest.rb#L235-L265) more or less for years.

Reply:
@juliantrueflynn if you do want to clean the dir entirely, I think you could use the [clean webpack plugin](https://github.com/johnagan/clean-webpack-plugin)

Reply:
@sbeam I think for rails/webpacker, that can work. However, I generally prefer that the webpack config be solely focused on webpack builds and not things like deleting files from some directory. In other words, I find plain shell commands more descriptive. Another issue is that many React + React + SSR projects build the client and server bundles in the same directory. So some thought must be given the order of the webpack builds, if concurrency is used, and ensuring that the clean stuff is configured right.